contracts/external/cw-abc/src/msg.rs

```diff
@@ -0,0 +1,185 @@
+use cosmwasm_schema::{cw_serde, QueryResponses};
+use cosmwasm_std::{Addr, Decimal, Decimal as StdDecimal, Uint128};
+
+use crate::abc::{CommonsPhase, CommonsPhaseConfig, CurveType, MinMax, ReserveToken, SupplyToken};
+
+#[cw_serde]
+pub struct InstantiateMsg {
+    /// The recipient for any fees collected from bonding curve operation
+    pub fees_recipient: String,
+
+    /// The code id of the cw-tokenfactory-issuer contract
+    pub token_issuer_code_id: u64,
+
+    /// Supply token information
+    pub supply: SupplyToken,
+
+    /// Reserve token information
+    pub reserve: ReserveToken,
+
+    /// Curve type for this contract
+    pub curve_type: CurveType,
+
+    /// Hatch configuration information
+    pub phase_config: CommonsPhaseConfig,
+
+    /// TODO different ways of doing this, for example DAO members?
+    /// Using a whitelist contract? Merkle tree?
+    /// Hatcher allowlist
+    pub hatcher_allowlist: Option<Vec<String>>,
+}
+
+/// Update the phase configurations.
+/// These can only be called by the owner.
+#[cw_serde]
+pub enum UpdatePhaseConfigMsg {
+    /// Update the hatch phase configuration
+    Hatch {
+        contribution_limits: Option<MinMax>,
+        // TODO what is the minimum used for?
```

Comment:
I think paired with an expiration, the minimum could be used to determine if the ABC was successful. 

If expired and the minimum was not met, then the ABC goes into the Closed phase. If expired with the minimum met, then the ABC goes into the Open phase.